src/test/java/simplifier/UserServiceImplTest.java

```diff
@@ -17,28 +10,57 @@
 import simplifier.repositories.UserRepository;
 import simplifier.services.UserServiceImpl;
 
+import java.util.Optional;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
 @RunWith(MockitoJUnitRunner.class)
 public class UserServiceImplTest {
 
-  @Mock
-  private UserRepository userRepository;
+    @Mock
+    private UserRepository userRepository;
+
+    @Mock
+    private BCryptPasswordEncoder bCryptPasswordEncoder;
+
+    @InjectMocks
+    private UserServiceImpl userService;
+
+    @Test
+    public void saveUser() {
+        User user = new User();
+
+        when(userRepository.save(user)).thenReturn(user);
+
+        Optional<User> savedUser = userService.saveUser(user);
+
+        assertThat(savedUser, is(Optional.of(user)));
 
-  @Mock
-  private BCryptPasswordEncoder bCryptPasswordEncoder;
+        verify(userRepository).save(user);
+        verify(bCryptPasswordEncoder).encode(user.getPassword());
+    }
 
-  @InjectMocks
-  private UserServiceImpl userService;
+    @Test
+    public void doNotSaveIfUsernameDuplicate() {
+        User existingUser  = new User();
+        existingUser.setUsername("test");
+        existingUser.setPassword("test");
 
-  @Test
-  public void saveUser() {
-    User user = new User();
+        User invalidUser = new User();
+        invalidUser.setUsername("test");
+        invalidUser.setPassword("test");
 
-    when(userRepository.save(user)).thenReturn(user);
+        when(userRepository.findByUsername(invalidUser.getUsername())).thenReturn(existingUser);
 
-    User savedUser = userService.saveUser(user);
+        Optional<User> savedUser = userService.saveUser(invalidUser);
 
-    assertThat(savedUser, is(equalTo(user)));
+        assertThat(savedUser, is(Optional.empty()));
 
-    verify(userRepository).save(user);
-  }
+        verify(userRepository).findByUsername(invalidUser.getUsername());
```

Comment:
Add verification that no more invocations of this mock are executed. It is possible to use save() method for example and test will pass, same is for bCryptPasswordEncoder. Use verifyNoMoreInteractions() in all tests or use should().then() instead of verify()